### DISCUSSION TITLE
What's the proper way to deal with the browser back button in SPA with unpoly?

### DISCUSSION THREAD

**Discussion by newk5** [unpoly/unpoly#326] [Q&A] [answered]
So I have a template in my SPA app (index.html) which looks a bit like this:
```html
<!DOCTYPE html>
<html lang="pt" >
    <head>
         imports for axios, unpoly, bootstrap, etc...
    </head>
    <body>
     logo image, etc...
    
      <div class="content">
         main page content here
      </div>
      
      <footer code here>
    </body>
 </html>   
````
Inside that ".content" div I have several ```<a>``` with up-target=".content" and with up-history set to true, they make calls to my server endpoints which return small HTML fragments and changes the contents of that div. I have configured URL rewriting on my server such that when any path/page of my app is requested, I rewrite the request to serve my template (index.html) and let my client side router (which is on index.html) render the page based on whats on the URL. This approach works fine for page refreshes, however it does not work when the user hits the browser back button. 

When I click the browser back button, it seems that the last request made on the previous "page" with unpoly is "re-played" (up:request:load event is triggered), in my case this is a request I made with up.render. The problem is, this last request only contains a small HTML fragment (page body which goes into that ".content div) and not the whole template, which results in the page incorrectly being rendered without the template. What's the "unpoly way" of dealing with this? The fact that unpoly replays the last request I made on the page that I'm going back to, is correct, however I'ld like to keep my whole template there when I press the back button and just have the unpoly request update the ".content" div just like when I refresh the page.

**Comment by adam12** [comment]
First off, are you using Unpoly 2? I believe Unpoly 1 had a bit harder time remembering where fragments were loaded from, which might explain some of the inconsistencies you seen. Unpoly 2 was a large refactor to improve this.

Second, can you share your Unpoly log? You can enable it using `up.log.enable()` in the browser console.

I always return full responses, so I'm not sure how much help I can be here, but at least with those 2 pieces of information perhaps we can build a bigger picture of what might be going wrong.

If you're using `up.render`, we might need to see that too.

**Reply by newk5** [reply]
Yep I'm using unpoly 2.4.1. I enabled the logs, here's a step by step breakdown of my situation:

- Users acesses index.html which at this point is a login page, user presses login button and up.render is called like this:

```javascript
  up.render({
        target: '.content',
        url: "/api/dashboard",
        history: true,
        method: "get",
        headers: {
            Authorization: 'secret token here',
        }
	});
```
This is the resulting log:

```
up:request:load Loading GET /api/dashboard
unpoly.min.js:1 up:request:loaded Server responded HTTP 200 to GET /api/dashboard (2120 characters)
unpoly.min.js:1 up:fragment:loaded Loaded fragment from successful response to GET /api/dashboard
unpoly.min.js:1 up.render() Updating ".content" in root layer
unpoly.min.js:1 up:location:changed Advanced to location /api/dashboard
unpoly.min.js:1 up.hello() Compiling fragment <div class=​"content" up-source=​"/​api/​dashboard">​…​</div>​
unpoly.min.js:1 up:fragment:inserted Inserted fragment <div class=​"content" up-source=​"/​api/​dashboard">​…​</div>​
```

- After the user is on the dashboard page he clicks:
```html
<a href="user/profile" up-target=".content" >Profile</a>
```
This is the resulting log:

```
up:link:follow Following link
unpoly.min.js:1 up:request:load Loading GET /api/user/profile
unpoly.min.js:1 up.request() Change with { solo } option will abort other requests
unpoly.min.js:1 up:request:loaded Server responded HTTP 200 to GET /api/user/profile (824 characters)
unpoly.min.js:1 up:fragment:loaded Loaded fragment from successful response to GET /api/user/profile
unpoly.min.js:1 up.render() Updating ".content" in root layer
unpoly.min.js:1 up:location:changed Advanced to location /api/user/profile
unpoly.min.js:1 up.hello() Compiling fragment <div class=​"content up-focusable-content" up-source=​"/​api/​user/profile" tabindex=​"-1">​…​</div>​
unpoly.min.js:1 up:fragment:inserted Inserted fragment <div class=​"content up-focusable-content" up-source=​"/​api/​user/profile" tabindex=​"-1">​…​</div>​

```
- Then the user clicks the browser back button and the initial call to /api/dashboard is called which only returns a fragment of HTML (not the full template)

This is the resulting log:

```
up:request:load Loading GET /api/dashboard
unpoly.min.js:1 up:request:loaded Server responded HTTP 200 to GET /api/dashboard (2120 characters)
unpoly.min.js:1 up:fragment:loaded Loaded fragment from successful response to GET /api/dashboard
unpoly.min.js:1 up.render() Updating "body" in root layer
unpoly.min.js:1 up.hello() Compiling fragment <body up-source=​"/​api/​dashboard">​…​</body>​
unpoly.min.js:1 up:fragment:inserted Inserted fragment <body up-source=​"/​api/​dashboard">​…​</body>​
unpoly.min.js:1 up.viewport.restoreScroll() Restoring scroll positions for URL /api/dashboard to {html: 0}
unpoly.min.js:1 up:location:changed Restored location /api/dashboard

```
At this point the page only contains the HTML fragment returned by the /api/dashboard endpoint and is incorrectly rendered since it's missing the whole template that was initially in index.html.

Anchor links also have the Authorization header and history set to true.

Unpoly seems to have no problem remembering that "/api/dashboard" was the last request made on the previous page, the problem is that it's assuming that the "/api/dashboard" endpoint is returning a full HTML page, which is wrong, since it only returns a small HTML fragment. Unpoly doesn't seem to remember that there was a full HTML page (template) on that previous page before the call to "/api/dashboard" was made.

**Reply by newk5** [reply]
Btw one thing I forgot to mention, my SPA does not use any URL hashes. Would this cause the wrong behaviour I mentioned above?

**Reply by adam12** [reply]
Can you create reproduction app using glitch template? Might help.

https://glitch.com/edit/#!/empty-unpoly-project

**Reply by newk5** [reply]
Hmm I dont think thats possible, since I'm doing calls to my backend, to reproduce it, I'd need to some backend that returns HTML, do you know of some I can use for testing?

**Reply by adam12** [reply]
I mocked up this, mimicking what you outlined. Does it seem correct?

https://glitch.com/edit/#!/shy-efficacious-drink

One issue out of the gate is that there is no history pushed for the Profile link. Not sure if this is intentional or not?

**Reply by newk5** [reply]
Sorry maybe I didn't explain properly, but as I said above, the Anchor links on my app have the history set to true too ( I added it globally using the "up : link:follow" event), so I am pushing history for the profile page too. I think I was able to reproduce the issue, check this one: https://glitch.com/edit/#!/powerful-polarized-hoverfly

Press login, then press profile, then press the browser back button. You'll see that the rest of the template disappears

**Reply by adam12** [reply]
That example was helpful and did demonstrate the issue.

I was able to fix it by configuring the restoreTarget to include that selector, but I'm really not sure if that's the correct way. Again, I always just return full responses so I'm not the best to speak to this.

Here's my remix with the config change and it working: https://glitch.com/edit/#!/tarry-three-angolatitan?path=scripts.js%3A4%3A47

Maybe Henning has a suggestion, as I know he does return partial responses in some of his apps.

**Reply by newk5** [reply]
Thank you, that worked perfectly. And yea I really like unpoly's concept of returning small HTML fragments instead of full HTML pages. I was curious of what was unpoly's way of dealing with that when the back buttton was used.

**Comment by triskweline** [comment]
> the problem is that it's assuming that the "/api/dashboard" endpoint is returning a full HTML page, which is wrong, since it only returns a small HTML fragment.

The server is expected to return HTML matching the `X-Up-Target` request header. When the user presses the back button Unpoly asks the user to respond for `X-Up-Target: body`. So your response must contain a full `<body>` element.

The server can choose to send smaller fragments, but only if Unpoly requested a smaller target. E.g. when the server sees request header `X-Up-Target: .sidebar` it can choose to only send a `<div class="sidebar">` element. It can also choose to send an entire HTML document which contains a `<div class="sidebar">` somewhere.

**Reply by newk5** [reply]
Thanks for the explanation I now understand why it happens, but I still have few questions if you dont mind 😛  

>  When the user presses the back button Unpoly asks the user to respond for X-Up-Target: body. So your response must contain a full <body> element`

Maybe I missunderstood, but doesn't this kinda defeat the whole point of having the server return small HTML fragments? If the user can press the back button at any time then, the server must always be prepared for it and always return full HTML pages in all requests, no? Or should the server have to check on every single request if ``X-Up-Target`` equals``body`` and respond accordingly? How do you usually deal with this when using unpoly in your projects?

> The server can choose to send smaller fragments, but only if Unpoly requested a smaller target

But why does unpoly not request the last up-target (in my case .content) when the user presses the back button and chooses to request the full ``X-Up-Target: body`` ?

**Comment by triskweline** [comment, marked as answer]
> How do you usually deal with this when using unpoly in your projects?

My default is to return full HTML pages for all requests, just like I would in a classic server-side web app with no JavaScript. Only if this is too slow (e.g. a large navigation that is expensive to render) I look at `X-Up-Target` and deliver a smaller fragment.

> If the user can press the back button at any time then, the server must always be prepared for it and always return full HTML pages in all requests, no? 

Yes. You also need to be prepared for this if the user goes to a URL directly, or presses the reload button. No Unpoly is involved here, but the browser will expect you to deliver a full HTML document.

**Reply by newk5** [reply]
> My default is to return full HTML pages for all requests, just like I would in a classic server-side web app with no JavaScript. Only if this is too slow (e.g. a large navigation that is expensive to render) I look at X-Up-Target and deliver a smaller fragment.

I see.

> You also need to be prepared for this if the user goes to a URL directly, or presses the reload button. No Unpoly is involved here, but the browser will expect you to deliver a full HTML document.

For these situations I use a client side router + URL rewriting on the server, this is how I solved it. Basically on the first full page load, I have some javascript that looks at the URL and depending on whats there, it calls up.render() with the corresponding server endpoint, which returns the small HTML fragment for the corresponding page. At that point the main template is already loaded so I just need the small html fragment. Unfortunately, the javascript that does these checks does not run when the user press the browser back button (probably due to bfcache) which is why I ran into my issue.

**Comment by triskweline** [comment]
> Maybe I missunderstood, but doesn't this kinda defeat the whole point of having the server return small HTML fragments? 

I think what you misread is: The point is not to render less HTML. This is only an optimization when rendering HTML is slow. The point is to only use a fragment of the server response, and discard the rest. This leads to much smoother navigation since we keep the state of the DOM we're not updating.

E.g. you can have a smaller interaction in a sidebar while keeping the main content. Or preserve focus and scroll positions in forms, because we're only updating a container with validation errors.

> But why does unpoly not request the last up-target (in my case .content) when the user presses the back button and chooses to request the full X-Up-Target: body ?

It's just not built this way currently. This could be an optimization worth looking into.

**Reply by newk5** [reply]
> I think what you misread is: The point is not to render less HTML. This is only an optimization when rendering HTML is slow. The point is to only use a fragment of the server response, and discard the rest

Yea I missunderstood that part, I thought the idea was small responses from the server, but I get your point now.
